src/pixel/homogeneous_pixel.rs

```diff
@@ -0,0 +1,125 @@
+use crate::*;
+
+/// A Pixel made up of a compile-time known number of contiguously stored color components and optionally an
+/// alpha component.
+///
+/// Unlike [`HeterogeneousPixel`] the alpha component must be the same type as the color
+/// components.
+///
+/// # Terminology
+///
+/// Component = An element of a pixel, inclusive of alpha. For example, [`Rgba`](crate::Rgba) is a pixel made up
+/// of four components, three color components and one alpha component.
+pub trait HomogeneousPixel:
+    HeterogeneousPixel<ColorComponent = Self::Component, AlphaComponent = Self::Component>
+{
+    /// The component type of the pixel used for both color and alpha components if any.
+    type Component: PixelComponent;
+
+    //TODO if const generic expressions become stable remove this associated type and just use
+    //`[Self::Component; Self::COMPONENT_COUNT]` as the return type.
+    //
+    /// The component array form of `Self`
+    type ComponentArray<R>: IntoIterator<Item = R> + AsSlice<R>;
+
+    /// Converts an owned `Pixel` type to an array of its components.
+    fn component_array(&self) -> Self::ComponentArray<Self::Component>;
+
+    /// Creates a new instance given an iterator of its components.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the iterator does not produce enough components for the pixel.
+    fn from_components(components: impl IntoIterator<Item = Self::Component>) -> Self;
```

Comment:
Isn't this redundant with `FromIterator`?

Reply:
Indeed, good spot! We should be able to add it as a trait bound.

Reply:
I'm unsure whether we should remove it and replace it with `FromIterator` which might be less discoverable for such a key function. Perhaps we leave it there but also add the `FromIterator` trait bound?

Reply:
But then from a simplicity standpoint we shouldn't need the `P::from_components()` method as it is duplication.

Reply:
I had in mind `impl FromIterator<C> for Pixel<C>`, so that `[1,2,3,4].into_iter().collect::<Rgba>()` works.

If libstd had a pixel type, it'd probably stick to `FromIterator` and things like `From<[C; 4]>`?

Discoverability could be solved with documentation? If the types implement all the standard methods,  `pixel.as_ref() == &[c; 4]`, then it would be discoverable in the sense that users could work with them as if they worked with `[t; 4]`, and simply expect all of the libstd methods they know to already work.

Reply:
Oh I see, via a blanket impl, what's the difference between a blanket impl: `impl FromIterator<C> for T: Pixel<C>` vs a super-trait bound: `pub trait Pixel: FromIterator<C>`?
I like the blanket version a bit more since it doesn't pollute the trait definition but I'm curious if there are any other differences.

Reply:
Okay when going to implement this I found that it is in fact *not* equivalent since `FromIterator` shouldn't panic whereas `from_components()` can/will panic.

Reply:
On that not perhaps we should change them to `try_` to give users more flexibility